File: etrike_protocol/profiles/xor8_ff_v1.py
"""Versioned XOR(bytes) ^ 0xFF integrity profile."""

from collections.abc import Iterable

PROFILE_ID = "xor8_ff_v1"


def compute(data: Iterable[int]) -> int:
    value = 0
    for byte in data:
        if not 0 <= byte <= 0xFF:
            raise ValueError("profile input bytes must be in range 0..255")
        value ^= byte
    return value ^ 0xFF


def verify(data: Iterable[int], checksum: int) -> bool:
    return 0 <= checksum <= 0xFF and compute(data) == checksum

File: autoware/src/our_packages/etrike_protocol/profiles/xor8_ff_v1.ts
/** Versioned XOR(bytes) ^ 0xFF integrity profile. */
export const PROFILE_ID = "xor8_ff_v1" as const;

export function compute(data: ArrayLike<number>): number {
  let value = 0;
  for (let index = 0; index < data.length; index += 1) {
    const byte = data[index];
    if (!Number.isInteger(byte) || byte < 0 || byte > 0xff) {
      throw new RangeError("profile input bytes must be integers in range 0..255");
    }
    value ^= byte;
  }
  return (value ^ 0xff) & 0xff;
}

export function verify(data: ArrayLike<number>, checksum: number): boolean {
  return Number.isInteger(checksum) && checksum >= 0 && checksum <= 0xff && compute(data) === checksum;
}
